pkg/sidecar-controller/framework_test.go

```diff
@@ -0,0 +1,1025 @@
+// 2) Call syncSnapshot/syncContent several times (both simulating "snapshot/content
+//    modified" events and periodic sync), until the controller settles down and
+//    does not modify anything.
+// 3) Simulate almost real API server/etcd and call add/update/delete
+//    content/snapshot.
+// In all these scenarios, when the test finishes, the framework can compare
+// resulting snapshots/contents with list of expected snapshots/contents and report
+// differences.
+
+// controllerTest contains a single controller test input.
+// Each test has initial set of contents and snapshots that are filled into the
+// controller before the test starts. The test then contains a reference to
+// function to call as the actual test. Available functions are:
+//   - testSyncSnapshot - calls syncSnapshot on the first snapshot in initialSnapshots.
+//   - testSyncSnapshotError - calls syncSnapshot on the first snapshot in initialSnapshots
+//                          and expects an error to be returned.
+//   - testSyncContent - calls syncContent on the first content in initialContents.
+//   - any custom function for specialized tests.
+// The test then contains list of contents/snapshots that are expected at the end
+// of the test and list of generated events.
+type controllerTest struct {
+	// Name of the test, for logging
+	name string
+	// Initial content of controller content cache.
+	initialContents []*crdv1.VolumeSnapshotContent
+	// Expected content of controller content cache at the end of the test.
+	expectedContents []*crdv1.VolumeSnapshotContent
+	// Initial content of controller Secret cache.
+	initialSecrets []*v1.Secret
+	// Expected events - any event with prefix will pass, we don't check full
+	// event message.
+	expectedEvents []string
+	// Errors to produce on matching action
+	errors []reactorError
+	// List of expected CSI Create snapshot calls
+	expectedCreateCalls []createCall
+	// List of expected CSI Delete snapshot calls
+	expectedDeleteCalls []deleteCall
+	// List of expected CSI list snapshot calls
+	expectedListCalls []listCall
+	// Function to call as the test.
+	test          testCall
+	expectSuccess bool
+}
+
+type testCall func(ctrl *csiSnapshotSideCarController, reactor *snapshotReactor, test controllerTest) error
+
+const testNamespace = "default"
+const mockDriverName = "csi-mock-plugin"
+
+var errVersionConflict = errors.New("VersionError")
+var nocontents []*crdv1.VolumeSnapshotContent
+var nosnapshots []*crdv1.VolumeSnapshot
+var noevents = []string{}
+var noerrors = []reactorError{}
+
+// snapshotReactor is a core.Reactor that simulates etcd and API server. It
+// stores:
+// - Latest version of snapshots contents saved by the controller.
+// - Queue of all saves (to simulate "content/snapshot updated" events). This queue
+//   contains all intermediate state of an object - e.g. a snapshot.VolumeName
+//   is updated first and snapshot.Phase second. This queue will then contain both
```

Comment:
This section needs to be updated.  It talks about snapshot.VolumeName, snapshot.Phase, etc.